File: C/merge-sort.c
#include <stdio.h>
#include "utils.h"
#include <stdbool.h>

#define N 10

void 
merge(int *arr, int l_idx, int m_idx, int r_idx, bool ascending)
{
    // define left array
    int left_len = m_idx - l_idx + 1;
    int left_arr[left_len];
    for (int i = 0; i < left_len; i++) {
        left_arr[i] = arr[l_idx + i];
    }

    // define right subarray
    int right_len = r_idx - (m_idx + 1) + 1;
    int right_arr[right_len];
    for (int i = 0; i < right_len; i++) {
        right_arr[i] = arr[m_idx + 1 + i];
    }

    // merge the two subarrays into the original one
    // while mantaining the starting range
    int i = 0;
    int j = 0;
    int k = l_idx;

    // while both subarrays have elements
    while (i < left_len && j < right_len) { 
        if (
            (ascending && left_arr[i] <= right_arr[j]) || 
            (!ascending && left_arr[i] > right_arr[j])
        ){
            arr[k++] = left_arr[i++];
        } else {
            arr[k++] = right_arr[j++];
        }
    }  
    
    // complete pushing the remaining element if any
    while (i < left_len) {
        arr[k++] = left_arr[i++];
    }

    while (j < right_len) {
        arr[k++] = right_arr[j++];
    }
}

void
merge_sort(int *arr, int l_idx, int r_idx, bool ascending)
{
    if (r_idx - l_idx == 0) return;

    int m_idx = (r_idx - l_idx) / 2 + l_idx;
    merge_sort(arr, l_idx, m_idx, ascending);
    merge_sort(arr, m_idx + 1, r_idx, ascending);
    merge(arr, l_idx, m_idx, r_idx, ascending);
}

int 
main(void)
{   
    bool asc_order = true;
    bool desc_order = !asc_order;

    int arr[N] = {2, 3, -3, 6, 5, -5, 10, 32, 4, 8};
    printf("PROBLEM INPUT:\n");
    print_array(arr, N);

    printf("\n");
    
    printf("INPUT SORTED (ASC order):\n");
    merge_sort(arr, 0, N - 1, asc_order); 
    print_array(arr, N);

    printf("\n");

    printf("INPUT SORTED (DESC order):\n");
    merge_sort(arr, 0, N - 1, desc_order);
    print_array(arr, N);
}

File: python/merge-sort.py
from typing import List
from math import floor

def merge (
    arr: List[int],
    l_idx: int,
    m_idx: int,
    r_idx,
    ascending: bool = True
) -> None:
    l_len = m_idx - l_idx + 1
    l_arr = [arr[i] for i in range(l_idx, m_idx + 1)]

    r_len = r_idx - (m_idx + 1) + 1
    r_arr = [arr[i] for i in range(m_idx + 1, r_idx + 1)]

    temp = None

    i, j, k = 0, 0, l_idx
    while i < l_len and j < r_len:
        if (
            (ascending and l_arr[i] <= r_arr[j]) or
            (not ascending and l_arr[i] > r_arr[j])
        ):
            arr[k] = l_arr[i]
            i += 1
            k += 1
        else:
            arr[k] = r_arr[j]
            j += 1
            k += 1

    while i < l_len:
        arr[k] = l_arr[i]
        i += 1
        k += 1

    while j < r_len:
        arr[k] = r_arr[j]
        j += 1
        k += 1



def merge_sort (
    arr: List[int],
    l_idx: int,
    r_idx: int,
    ascending: bool = True
) -> None:
    if (r_idx - l_idx == 0):
        return
    m_idx = floor((r_idx + l_idx) / 2)
    merge_sort(arr, l_idx, m_idx, ascending)
    merge_sort(arr, m_idx + 1, r_idx, ascending)
    merge(arr, l_idx, m_idx, r_idx, ascending)


if __name__ == "__main__":
    desc_order = False

    arr = [2,5,3,7,4,8,-1,-6,10,7]
    print("PROBLEM INPUT:")
    print(f"arr{arr}")
    print()

    merge_sort(arr, 0, len(arr) - 1)
    print("INPUT SORTED (ASC order):")
    print(f"arr{arr}")
    print()

    merge_sort(arr, 0, len(arr) -1, desc_order)
    print("INPUT SORTED (DESC order):")
    print(f"arr{arr}")
    print()
        
